### DISCUSSION TITLE
v3 component architecture decisions

### DISCUSSION THREAD

**Discussion by veloware** [chakra-ui/chakra-ui#8327] [General]
I'm currently building a component library using Chakra-UI as a base, and I find myself spending a lot of time digging through Chakra-UI source for inspiration on how to best structure components of my own, in particular complex / compound components.

One of the patterns Ive been using which is somewhat inspired by how components are built in Chakra, and in general, just following a pattern of 'composition over configuration'.

One example I have built is a `Tile` component, similar to a `Card` in Chakra-UI.  Ive been taking a slightly different approach in how I have the consumer compose their components in that I am using props to pass components, whereas I noticed in Chakra, the various anatomy components that build up the `Card` component seem to all being passed as Children, e.g.

```
<Card>
  <CardHeader>
    <Heading size='md'>Client Report</Heading>
  </CardHeader>
  <CardBody>
     // content
  </CardBody>
  <CardFooter>
     // footer content
  </CardFooter>
</Card>
```

whereas I have being doing more like - 

```
<Tile
   header={
       <TileHeader></TileHeader>
   }
   footer={
       <TileFooter></TileFooter>
   }
>
   // body content (children)
</Tile>
```

And I know that Chakra used to use this approach, for example, with the Button component and the `leftIcon` prop, however I see in the v3 migration guide, it seems to be moved away from this prop-based approach and everything is just `children`

From the Docs - 

> leftIcon and rightIcon
> 
> Removed leftIcon and rightIcon in favor of rendering an icon component inlined with the button content.
> 
> To implement iconSpacing, you can use the gap prop on the Button component.
> 
> Before:
> 
> ```
> <Button leftIcon={<AddIcon />}>Click me</Button>
> ```
> 
> After:
> 
> ```
> <Button>
>   <AddIcon />
>   Click me
> </Button>
> ```

Another similar, but slightly different example is the Avatar component, from the v3 migration docs - 

> Avatar
> Before:
> 
> ```
> <Avatar name="Christian Nwamba" src="https://bit.ly/code-beast" />
> ```
> 
> After:
> 
> ```
> <Avatar.Root name="Dan Abrahmov" src="https://bit.ly/dan-abramov">
>   <Avatar.Image />
>   <Avatar.Fallback />
> </Avatar.Root>
> ```

For the Avatar component we pass the same props into the root element, but overall the footprint of the code that the consumer has to write has increased, there are many other similar examples in the migration docs.

And a final example, the Progress component, from the migration docs - 

> Progress
> Before:
> 
> ```
> <Progress value={50} />
> ```
> 
> After:
> 
> ```
> <Progress.Root value={50}>
>   <Progress.Track>
>     <Progress.FilledTrack />
>   </Progress.Track>
>   <Progress.ValueText />
> </Progress.Root>
> ```

From what I can see, for the Progress component, again each part of the component is exposed and has to be composed up by the consumer, in this case it seems like its at almost every level of the DOM tree has been exposed and composed, yet each of those components would be required to make a Progress bar make sense?

Im curious to understand the reasoning of not having the parent component control the layout of its child components? It seems from looking at a lot of the v3 components, `children` seems to be the go-forward approach, I would like to apply the same approach to my library to align with Chakra, but would like to understand the reasoning of this design decision?  Essentially the question I want to answer in my head is - 

When would I use props over children, and vice versa?

I can see how this works when the layout inside the component is quite linear, either vertically or horizontally, but when the layout of a component is 2 dimensional, I think individual props would be needed to define those 'slots' for the child components to be placed into?

Thanks!

**Comment by segunadebayo** [comment]
v3 is slots as children. That is the model.

Use snippets when you want the composed version in one file.
